remote_attestation/rust/src/attestation.rs

```diff
@@ -15,238 +15,545 @@
 //
 
 use crate::{
-    crypto::{get_sha256, AeadEncryptor, EncryptorType, KeyNegotiator},
-    proto::{AttestationIdentity, AttestationInfo, AttestationReport},
+    crypto::{
+        get_random, get_sha256, AeadEncryptor, KeyNegotiator, KeyNegotiatorType, SignatureVerifier,
+        Signer,
+    },
+    proto::{AttestationInfo, AttestationInit, AttestationReport, ClientIdentity, ServerIdentity},
 };
 use anyhow::{anyhow, Context};
 
-/// Remote attestation protocol implementation.
-/// Template `S` defines the type of attestation used. It can be one of:
-/// - [`UnattestedSelf`]: represents an attestation process, where current machine remotely attests
-///   to a remote peer and sends attestation info to it.
-/// - [`UnattestedPeer`]: represents an attestation process, where current machine remotely attests
-///   a remote peer and verifies its attestation info.
-/// - [`MutuallyUnattested`]: represents an attestation process, where current machine and a remote
-///   peer remotely attest each other.
-pub struct AttestationEngine<S>
+const ATTESTATION_PROTOCOL_VERSION: u8 = 1;
+/// Size (in bytes) of a random array sent in messages to prevent replay attacks.
+const REPLAY_PROTECTION_ARRAY_SIZE_BYTES: usize = 32;
+
+/// Convenience trait for transitioning between states in the state machine.
+pub trait StateMachine<T, A> {
+    fn transition(self, argument: A) -> anyhow::Result<T>;
+}
+
+impl<P> StateMachine<AttestationEngine<P, Attesting>, KeyNegotiator>
+    for AttestationEngine<P, Initializing>
 where
-    S: AttestationType,
+    P: AttestationParticipant,
 {
-    /// Type of the remote attestation process.
-    attestation_type: S,
-    /// Implementation of the X25519 Elliptic Curve Diffie-Hellman (ECDH) key negotiation.
-    key_negotiator: KeyNegotiator,
+    fn transition(
+        self,
+        key_negotiator: KeyNegotiator,
+    ) -> anyhow::Result<AttestationEngine<P, Attesting>> {
```

Comment:
I don't fully see the point of the generic type here; in what case would you ever use an AttestationEngine without caring about the specific state it is in? It seems to me these may as well be completely distinct types, e.g. AttestationEngineInit, AttestationEngineAttesting, etc. (probably with better naming than the ones I am suggesting here), and explicit methods to move from one to the other. In particular, is there any functionality that is common to all the states (apart from the ability to move to the next state)? I'm probably missing something though, so please feel free to correct me :)

Reply:
Empty classes were added more for readability.
Because if I remove `participant` and `state` classes, then the code will contain:
- `ClientAttestationEngineInit`
- `ClientAttestationEngineAttesting`
- `ClientAttestationEngineAttested`
- `ServerAttestationEngineInit`
- `ServerAttestationEngineAttesting`
- `ServerAttestationEngineAttested`

And having `AttestationEngine<Participant, State>` seems a bit nicer.

Also making `AttestationEngine` a parent class is currently not possible, because each `<Participant, State>` returns a separate Protobuf message. So processing handshake messages is basically a new function everytime (It could be updated, but only after implementing https://github.com/project-oak/oak/issues/2273, when all responses will look like `Vec<u8>`).

Reply:
I am also no longer entirely sure what benefit we are getting from using the state machine pattern. Is it readability, or correctness or compile-time error checking? I guess it will be clearer after #2273.

Reply:
`Server` and `Client` have been removed.
Currently there are sepatate `ClientAttestationEngine` and `ClientAttestationEngine` types.